File: src/wen_mscjs.cpp
#include <TMB.hpp>



template<class Type>
struct pim: vector<matrix<int> > {

  pim(SEXP x){ // Constructor
    (*this).resize(LENGTH(x));
    for(int i=0; i<LENGTH(x); i++){
      SEXP m = VECTOR_ELT(x, i);
    (*this)(i) = asMatrix<int>(m);
    }

  }
};



template<class Type>
Type objective_function<Type>::operator() ()
{
//~~~~~~~~~~~~~~~~~~~
// Data
//~~~~~~~~~~~~~~~~~~~
DATA_INTEGER(n_OCC);      //number of total survival/ reacputure occasions;
DATA_INTEGER(nDS_OCC);      //number of downstream survival/ reacputure
DATA_INTEGER(n_states);      //number of possible adult return ages (i.e. statres -3) 
DATA_INTEGER(n_groups);      //number groups (i.e., unique combos of LH,stream,downstream,year). Used in psi mlogit backtransform
DATA_INTEGER(n_unique_CH); //number of unique capture occasions
DATA_IMATRIX(CH);           //capture histories (excluding occasion at marking (which we are conditioning on))
DATA_IVECTOR(freq);       //frequency of capture histories
DATA_MATRIX(X_phi);         //fixed effect design matrix for phi
DATA_MATRIX(X_p);         // fixed effect design matrix for psi
DATA_MATRIX(X_psi);         // fixed effect design matrix for psi
DATA_STRUCT(Phi_pim, pim); //index vector of matrices for phi parameter vector for a given Ch x occasion
DATA_STRUCT(p_pim, pim);   //index vector of matrices for p parameter vector for a given Ch x occasion
DATA_IVECTOR(Psi_pim);
DATA_IVECTOR(fix_p_last);
  //~~~~~~~~~~~~~~~~~~~
  // Parameters
  //~~~~~~~~~~~~~~~~~~~
  PARAMETER_VECTOR(beta_phi);  //Phi fixed effect coefficients
  PARAMETER_VECTOR(beta_p);    //p fixed effect coefficients
  PARAMETER_VECTOR(beta_psi);    //psi fixed effect coefficients

  //~~~~~~~~~~~~~~~~~~~
  // Variables
  //~~~~~~~~~~~~~~~~~~~
  
  // Joint negative log-likelihood
  parallel_accumulator<Type> jnll(this);

  // Linear predictors
  vector<Type> eta_phi = X_phi*beta_phi;
  vector<Type> eta_p = X_p*beta_p;
  vector<Type> eta_psi = X_psi*beta_psi;


  // Apply link
  vector<Type> phi=invlogit(eta_phi);
  vector<Type> p=invlogit(eta_p);
  for (int i =0; i<fix_p_last.size(); i++) p(fix_p_last(i))=1; //fix detection at 1 on last occasion
  //** will this crash things if fix_p_last.size() =0?**
  REPORT(phi);
  REPORT(p);
  ////phi inverse multinomial logit
  matrix<Type> psi(n_groups,n_states);
  eta_psi= exp(eta_psi);
  vector<Type> denom = eta_psi.segment(0,n_groups)+eta_psi.segment(n_groups,n_groups)+Type(1);
  psi.col(0)= Type(1)/denom;                      //return after 1 year
  psi.col(1)= eta_psi.segment(0,n_groups)/denom; //return after 2 year
  psi.col(2)= eta_psi.segment(n_groups,n_groups)/denom; //return after 3 year
  REPORT(psi);

  //~~~~~~~~~~~~~~~~~~~
  // Likelihood
  //~~~~~~~~~~~~~~~~~~~

  ////Variables
  vector<Type> pS(4); //state probs: dead, 1, 2, 3
  Type u = 0;         // holds the sum of probs after each occasion
  Type NLL_it=0;      // holds the NLL for each CH
  Type tmp = 0;       // holds the prob of a given state during observation process in upstream migration

  for(int n=0; n<n_unique_CH; n++){ // loop over individual unique capture histories
    pS.setZero(); //initialize at 0,1,0,0 (conditioning at capture)
    pS(1)=Type(1);
    NLL_it=Type(0); //initialize capture history NLL at 0

  //downstream migration
  for(int t=0; t<nDS_OCC; t++){       //loop over downstream occasions (excluding capture occasion)
    //survival process
    pS(0) += Type((Type(1)-phi(Phi_pim(0)(n,t)))*pS(1)); //prob die or stay dead
    pS(1) *= Type(phi(Phi_pim(0)(n,t))); //prob stay alive
    //observation process
    pS(1) *= Type(p(p_pim(0)(n,t))*CH(n,t)+ (Type(1)-p(p_pim(0)(n,t)))*(Type(1)-CH(n,t))); //prob observation given alive
    pS(0) *= Type(Type(1)-CH(n,t)); //prob observation given dead
    //acculate NLL
    u = pS.sum();  //sum of probs
    pS = pS/u; //normalize probs
    NLL_it  +=log(u);    //accumulate nll
  }

  //ocean occasion
  int t = nDS_OCC;  //set occastion to be ocean occasion
  ////survival process
  pS(0) += Type((Type(1)-phi(Phi_pim(0)(n,t)))*pS(1)); //prob die or stay dead in ocean
  pS(1) *= Type(phi(Phi_pim(0)(n,t))); //prob survive ocean
  //maturation age process
  pS(2) = pS(1) * psi(Psi_pim(n),1);
  pS(3) = pS(1) * psi(Psi_pim(n),2);
  pS(1) *= psi(Psi_pim(n),0);
  ////observation process
  if(!CH(n,t)){
  pS(1) *= Type(Type(1)-p(p_pim(0)(n,t)));
  pS(2) *= Type(Type(1)-p(p_pim(1)(n,t)));
  pS(3) *= Type(Type(1)-p(p_pim(2)(n,t)));
  } else{
    tmp=pS(CH(n,t))*p(p_pim((CH(n,t)-1))(n,t));
    pS.setZero();
    pS(CH(n,t))=tmp;
  }
  //acculate NLL
  u = pS.sum();  //sum of probs
  pS = pS/u; //normalize probs
  NLL_it  +=log(u);    //accumulate nll
  //end ocean occasion

  //upstream migration
  for(int t=(nDS_OCC+1); t<n_OCC; t++){       //loop over upstream occasions
    ////survival process
    pS(0) += Type((Type(1)-phi(Phi_pim(0)(n,t)))*pS(1))+
      Type((Type(1)-phi(Phi_pim(1)(n,t)))*pS(2))+
      Type((Type(1)-phi(Phi_pim(2)(n,t)))*pS(3));  // sum(prob vec * 1, 1-phi_1, 1-phi_2, 1-phi_3)
    pS(1) *= Type(phi(Phi_pim(0)(n,t)));                 // sum(prob vec * 0,   phi_1,       0,       0)
    pS(2) *=  Type(phi(Phi_pim(1)(n,t)));                 // sum(prob vec * 0,       0,   phi_2,       0)
    pS(3) *=  Type(phi(Phi_pim(2)(n,t)));                 // sum(prob vec * 0,       0,       0,   phi_3)
    ////observation process
    if(!CH(n,t)){
      pS(1) *=  Type((Type(1)-p(p_pim(0)(n,t))));
      pS(2) *=  Type((Type(1)-p(p_pim(1)(n,t))));
      pS(3) *=  Type((Type(1)-p(p_pim(2)(n,t))));
    }else{
      tmp=pS(CH(n,t))*p(p_pim((CH(n,t)-1))(n,t));
      pS.setZero();
      pS(CH(n,t))=tmp;
    }
    //acculate NLL
  u = pS.sum();  //sum of probs
  pS = pS/u; //normalize probs
  NLL_it  +=log(u);    //accumulate nll
  }
  //multiply the NLL of an individual CH by the frequency of that CH and subtract from total jnll
  jnll-=(NLL_it*freq(n));
  }
  //end of likelihood
  //return jnll
  return(jnll);
}

File: src/Archive/Wen_MSCJS.R
library(here)
library(tidyverse)
library(TMB)
source(here("src","data_proc.r"))

#subset columns needed for analysis
dat_out<-select(mark_file_CH,sea_Year_p,LH,stream, #grouping variables
                McN_J,Bon_J,Bon_A,Tum_A) %>% #sites/occasions to include in model
  #first year with all stream data through last year where data on all three return ages is available (because it is 2020)
  filter(sea_Year_p>=2007&sea_Year_p<=2017) %>%
  #make grouping variables factors
  mutate_at(vars(sea_Year_p:stream),~as.factor(as.character(.x))) %>% 
  #create multistate capture histories
  mutate(ch=select(., McN_J:Tum_A) %>%  reduce(paste0)) %>%
  #mutate(ch=select(., McN_J,Bon_J,Est_J) %>% reduce(paste0)) %>%
  mutate(ch=paste0("1",ch)) %>% 
  #reduce data to unqiue capture history/ groups combos and counts
  group_by_all() %>% summarise(freq=n()) %>% as.data.frame()


#get number of site/occasions
nOCC<-nchar(dat_out$ch[1])-1
#get number of downstream sites/occasiosn
nDS_OCC<-sum(substr(colnames(dat_out %>% select( McN_J:Tum_A)),5,5)=="J")
#numbr of unique capture histories
n_unique_CH<-nrow(dat_out)
#number of states
n_states<-3

#process data using RMark function. Specifies grouping variables for parameters, and type of model and hence parameters. "Multistrate" used S(Phi), p, and psi
wenatchee.processed<-RMark::process.data(dat_out,model="Multistrata",groups=c("LH","stream","sea_Year_p"))

#Make design data using RMark function. Sets up matrices for each parameter where there is a row for each combination of the grouping variables, occasion, age, etc. Specifiying the pim.type can reduce some of the combinations/# of rows. See?RMark::make.design.data() for more info. 
wenatchee.ddl<-RMark::make.design.data(wenatchee.processed,parameters=list(S=list(pim.type="time"), #survival changes by occasion and potentially stratum(i.e. fish age)
                                                                           p=list(pim.type="time"),#detection changes by occasion and potentially stratum(i.e. fish age)
                                                                           Psi=list(pim.type="constant")), # state (age at return) transitions only occur at one time period.
                                       remove.unused = TRUE)


#extract the individual design data for each parameter
##phi/s survival
Phi.design.dat<-wenatchee.ddl$S %>% 
  filter(Time>(nDS_OCC-1)|stratum==1) #cant be in strata (fish age) other than 1 on downstream, or ocean for survival. Note "Time" column indexing starts at 0, so occasion/Time (nDS_OCC-1) is the last juvenile detection occasion

##p detection, same as Phi
p.design.dat<-wenatchee.ddl$p %>% 
  filter(Time>(nDS_OCC-1)|stratum==1)# can be in multiple strate for detection at time (nDS_OCC) but nor for survival. Will fix detection at last time to 1 down below when nitializing model.

##Psi transition. 
Psi.design.dat<-wenatchee.ddl$Psi %>% 
  filter(stratum==1) %>%  #Can only transition from state 1 (Juvenile entering ocean)
  arrange(tostratum,group) # sort by stratum and group so the first half of rows represents the alr probs of transitioning to age 2 and the second half of the rows the alr probs of transitioning to state 3. This is neccesary for the way I am coding this in TMB, to take use the two halves of the vectors when doing the backtransformation from alr to simplex. 

#make PIMs, which are matrices (nCH by nTimes) that give the index of the paramater wiithin the vector of phi or p parameters

##Phi pim. list, where each element is a PIM for a different state
###list of empty matrices (nCH x nOCC)
Phi_pim<-rep(list(matrix(NA,n_unique_CH,nOCC)),n_states)
####fill in pim matrix for Phi (survival)
for ( i in 1:nDS_OCC){#loop over downstream "times" (e.g. McN_j, Bon_j, Est_j)
  #fill in matrix for state 1 (only possible state for downstream migration)
  Phi_pim[[1]][,i]<-match(paste0(dat_out %>% select(LH,stream,sea_Year_p) %>% reduce(paste0),"time",i), #match group (LH,stream, seaward migration year) and occasion (time) for each CH
                          paste0(Phi.design.dat$group,"time",Phi.design.dat$time))-1 #with corresponding row in design data (which will become corresponding element of parameter vector). subtract 1 becauuse TMB indexing starts at 0
}

for ( i in (nDS_OCC+1):nOCC){ #loop over upstream "times"/occasions
  for(j in 1:n_states) #loop over "states" (i.e. age of fish/years at sea)
    Phi_pim[[j]][,i]<-match(paste0(dat_out %>% select(LH,stream,sea_Year_p) %>% reduce(paste0),"time",i,"stratum",j), #match group (LH,stream, seaward migration year), occasion (time), and stratum (fish age/return year) for each CH
                            paste0(Phi.design.dat$group,"time",Phi.design.dat$time,"stratum",Phi.design.dat$stratum))-1#with corresponding row in design data (which will become corresponding element of parameter vector)
}

##phi_pim and p_pim are the same. There is a detection and survival for each occasion, although I will fix detection at the last occasion below.
p_pim<-Phi_pim

##Psi pim. (nCH length vector) for each CH prob of row of ALR prob matrix of transition to states 2 or 3 (columns 1 or 2) corresponding to returning to Bonneville as adults after 2 or 3 years. The design data has been ordered such that the first half is for transition to state 2 and the second hallf for transition to state 3, so I only need the index corrsponding with the transitin to state 2 in the design data matrix
Psi_pim<-match(paste0(dat_out %>% select(LH,stream,sea_Year_p) %>% reduce(paste0),"tostratum",2),
               paste0(Psi.design.dat$group,"tostratum",Psi.design.dat$tostratum))-1 #subtract 1 becauuse TMB indexing starts at 0

#number of groups in the psi deisgn matrix
n_groups<-nrow(Psi.design.dat)/2

#glmmTMB objects to get design matrices etc. for each parameter
## phi
Phi.design.glmmTMB<-glmmTMB::glmmTMB(par.index~-1+time,data=Phi.design.dat,dispformula = ~0,doFit=FALSE)
## p
p.design.glmmTMB<-glmmTMB::glmmTMB(par.index~-1+time,data=p.design.dat,dispformula = ~0,doFit=FALSE)

## psi
Psi.design.glmmTMB<-glmmTMB::glmmTMB(par.index~-1+tostratum,data=Psi.design.dat,dispformula = ~0,doFit=FALSE)

setwd(here("Src"))
#compile and load TMB model
TMB::compile("wen_mscjs.cpp")
dyn.load(dynlib("wen_mscjs"))

#set detection at the last site to be very high. I will fix this value, because otherwise it is unidentifiable, and other analyses  (i.e. looking at detection of adult fish that were picked up on instream arrays and whether they were detected at Tumwate) suggests detestion at Tumwater Dam IS 100%

#Make data for TMB
dat_TMB<-list(
  n_OCC=nOCC,
  nDS_OCC=nDS_OCC,
  n_states=n_states,
  n_groups=n_groups,
  n_unique_CH=n_unique_CH,
  CH=select(dat_out,McN_J:Tum_A) %>% as.matrix(),
  freq=dat_out$freq,
  X_phi=Phi.design.glmmTMB$data.tmb$X,
  X_p=p.design.glmmTMB$data.tmb$X,
  X_psi=Psi.design.glmmTMB$data.tmb$X,
  Phi_pim=Phi_pim, #indexing starts at 0
  p_pim=p_pim,
  Psi_pim=Psi_pim,
  fix_p_last=numeric(0)
)


#make param inits for TMB
par_TMB<-list(
  beta_phi=Phi.design.glmmTMB$parameters$beta,
  beta_p=p.design.glmmTMB$parameters$beta,
  beta_psi=Psi.design.glmmTMB$parameters$beta
)  


#map to fix parameter
map<-list(
  beta_p=factor(seq(length(par_TMB$beta_p )))
  #  beta_psi=factor(c(NA,NA))
)
#fix detection prob at last time at 1
map$beta_p[substr(colnames(dat_TMB$X_p),5,5)==(nOCC+1)]<-NA #last occastion detection is mapped
map$beta_p<-map$beta_p %>% droplevels()
fix_p_last<-which(p.design.dat$time==(nOCC+1))-1 # fix lest occasion detection at 1 (see TMB model .cpp file for details)
dat_TMB$fix_p_last<-fix_p_last

#initialize model
mod<-TMB::MakeADFun(data=dat_TMB,parameters = par_TMB,map=map,DLL ="wen_mscjs")
#fit model
for ( i in 1:3){
  fit<-nlminb(mod$env$last.par.best,mod$fn,mod$gr)
  }
fit
#sd report
sd_rep<-sdreport(mod)
sd_rep
#report
mod_rep<-mod$report()
head(mod_rep$psi)
mod_rep$phi[1:4]
mod_rep$p[1:3]

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Mark model for comparison
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
dat_out<-select(mark_file_CH,sea_Year_p,LH,stream, #grouping variables
                McN_J,Bon_J,Bon_A,Tum_A) %>% #sites/occasions to include in model
  #first year with all stream data through last year where data on all three return ages is available (because it is 2020)
  filter(sea_Year_p>=2007&sea_Year_p<=2017) %>%
  #make grouping variables factors
  mutate_at(vars(sea_Year_p:stream),~as.factor(as.character(.x))) %>% 
  #create multistate capture histories
  mutate(ch=select(., McN_J:Tum_A) %>%  reduce(paste0)) %>%
  #mutate(ch=select(., McN_J,Bon_J,Est_J) %>% reduce(paste0)) %>%
  mutate(ch=paste0("1",ch)) 

wenatchee.processed=process.data(dat_out,model="Multistrata",groups=c("LH","stream"))
#wenatchee.processed=process.data(test_dat,model="MSCJS",strata.labels=c("A","B","C"))
wenatchee.ddl=make.design.data(wenatchee.processed)



#
# p
#
#
# States B&C dont exist for juveniles (times 2-5)

#wenatchee.ddl$p<-wenatchee.ddl$p[-wenatchee.ddl$p$Time<=2&wenatchee.ddl$p$stratum!="A",]
#wenatchee.ddl$S<-wenatchee.ddl$S[-wenatchee.ddl$S$Time<=2&wenatchee.ddl$S$stratum!="A",]



#column for migration year where onupstream state A is seaward year +1, b +2, and c +3
seaward_year_vec<-as.numeric(substr(wenatchee.ddl$p$sea_Year_p,2,5))
wenatchee.ddl$p$det_year<-factor(ifelse(wenatchee.ddl$p$Time<2,seaward_year_vec,ifelse(wenatchee.ddl$p$stratum==1,seaward_year_vec+1,ifelse(wenatchee.ddl$p$stratum==2,seaward_year_vec+2,seaward_year_vec+3))))



#wenatchee.ddl$p$fix=ifelse((wenatchee.ddl$p$stratum!="A"&wenatchee.ddl$p$Time<=4),0,NA)
# set detection at Tumwater adult at 100% based on other work
wenatchee.ddl$p$fix[wenatchee.ddl$p$time==5]=1


#
# Psi
#
# can only change state (other than death) at time 4
# rest are fixed values
wenatchee.ddl$Psi$fix=NA
# stay in current state (year) other than time 4 (ocean)
wenatchee.ddl$Psi$fix[wenatchee.ddl$Psi$stratum==
                        wenatchee.ddl$Psi$tostratum & wenatchee.ddl$Psi$time!=(3)]=1

wenatchee.ddl$Psi$fix[wenatchee.ddl$Psi$stratum!=
                        wenatchee.ddl$Psi$tostratum &wenatchee.ddl$Psi$time!=(3)]=0

wenatchee.ddl$Psi$fix[wenatchee.ddl$Psi$stratum!=1 &
                        wenatchee.ddl$Psi$time==3]=0



# 
# S
#
#Add a columna for whether first occasion, which is necessarily  different for subyearlings and yearlings
wenatchee.ddl$S$time_1<-ifelse(wenatchee.ddl$S$time==1,1,0)
#
#column for yearlings vs subs
wenatchee.ddl$S$Yrlng<-ifelse(wenatchee.ddl$S$LH=="smolt",1,0)
#
#columns for whether white river, which might be different because above lake Wenatchee
wenatchee.ddl$S$White<-ifelse(wenatchee.ddl$S$stream=="White",1,0)
#
#
#column for migration year where onupstream state A is seaward year +1, b +2, and c +3
seaward_year_vec<-as.numeric(substr(wenatchee.ddl$S$sea_Year_p,2,5))
wenatchee.ddl$S$det_year<-factor(ifelse(wenatchee.ddl$S$Time<2,seaward_year_vec,ifelse(wenatchee.ddl$S$stratum=="A",seaward_year_vec+1,ifelse(wenatchee.ddl$S$stratum=="B",seaward_year_vec+2,seaward_year_vec+3))))


wenatchee.ddl$S$upstream_time<-ifelse(wenatchee.ddl$S$Time>2,1,0)


# fixing S to 1 in strata where fish cant be
wenatchee.ddl$S$fix<-NA
wenatchee.ddl$S$fix[wenatchee.ddl$S$stratum!=1 &
                      wenatchee.ddl$S$Time<3]=1 #using numeric "Time" which starts at 0 hence 3 is 4





# fit model
p.timexstratum.tag=list(formula=~-1+time)
Psi.sxtime=list(formula=~-1+tostratum)
S.stratumxtime=list(formula=~-1+time)



test<-RMark::mark(wenatchee.processed,wenatchee.ddl,
                  model.parameters=list(S=S.stratumxtime,p= p.timexstratum.tag,Psi=Psi.sxtime),run=TRUE)

real_test<-get.real(test,par="p")
real_test$`Group:LHsmolt.streamChiwawa.sea_Year_py2007`

test$results$beta

sd_rep$value

#compare parameters
sd_rep$par.fixed - test$results$beta[,1]
#compare standard errors
sqrt(diag(sd_rep$cov.fixed)) - test$results$beta[,2]
